### PR TITLE
Adds LSE output for templated-attention-hop if inputs require grad

### PR DESCRIPTION
**Pull request by drisspg**
Adds LSE output for templated-attention-hop if inputs require grad

Prep PR for adding autograd support to templated-attention-hop. The kernel needs to output the LSE during the forward which will be used during backwards. 

### Output code
https://gist.github.com/drisspg/2aea3ce5db75811e7e143eeecb774d8a

## Before
| Type    |   Speedup |   batch_size |   num_heads |   q_seq_len |   k_seq_len |   head_dim | score_mod     | dtype          |
|---------|-----------|--------------|-------------|-------------|-------------|------------|---------------|----------------|
| Average |     1.159 |              |             |             |             |            |               |                |
| Max     |     1.342 |           16 |          16 |         512 |         512 |         64 | noop          | torch.bfloat16 |
| Min     |     1.016 |            1 |          16 |         512 |         512 |         64 | relative_bias | torch.bfloat16 |

## After
 Type    |   Speedup |   batch_size |   num_heads |   q_seq_len |   k_seq_len |   head_dim | score_mod   | dtype          |
|---------|-----------|--------------|-------------|-------------|-------------|------------|-------------|----------------|
| Average |     1.155 |              |             |             |             |            |             |                |
| Max     |     1.339 |           16 |          16 |         512 |         512 |         64 | noop        | torch.bfloat16 |
| Min     |     1.009 |            1 |          16 |         512 |         512 |         64 | head_bias   | torch.bfloat16 |

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -5737,7 +5745,7 @@ def create_placeholder(name: str, dtype: torch.dtype) -> InputBuffer:
             ]:
                 sdpa_template.maybe_append_choice(
                     choices=choices,
-                    input_nodes=(query, key, value),
+                    input_nodes=(query, key, value, logsumexp),
```
Comment (Chillee) [inline torch/_inductor/lowering.py:5748]: btw, we need some way for the template to say which one of the inputs are being mutated.

File: torch/_dynamo/variables/higher_order_ops.py [hunk 2]
```diff
@@ -1384,13 +1379,18 @@ def call_function(
             args, kwargs
         )
 
-        p_args, p_kwargs, example_value = self.create_wrapped_node(tx, query, score_mod)
+        p_args, p_kwargs = self.create_wrapped_node(tx, query, score_mod)
         proxied_args = [query, key, value, *other_buffers]
 
         # Store the invocation as a call
         # Norm_kwargs contains the score_function and we dont want to proxy this because
```
Comment (drisspg) [inline torch/_dynamo/variables/higher_order_ops.py:1386]: update the comment, the problem was that I copied pasta from the rest of the HOP implementations. However unlike the others, the subgrpah's output for this hop is unrelated to what the overall HOP returns, and thus needed to create the right output proxy.  I am not sure why this wasn't erroing before as the shapes for the output proxies were wonky.

File: torch/_inductor/kernel/templated_attention.py [hunk 3]
```diff
@@ -149,19 +149,21 @@ def sdpa_grid(batch_size, num_heads, num_queries, d_model, meta):
...
     idx_z = tl.program_id(1) // H
     idx_h = tl.program_id(1) % H
     idx_m = offs_m[:, None]
     idx_d = tl.arange(0, BLOCK_DMODEL)[None, :]
+
     # TODO generalize and add proper mask support
     mask = (idx_m != -1) & (idx_d != -1)
     {{store_output(("idx_z", "idx_h", "idx_m", "idx_d"), "acc")}}
+
+    # TODO dont want to write this if we dont require grad
```
Comment (drisspg) [inline torch/_inductor/kernel/templated_attention.py:163]: Will try this

File: test/inductor/test_templated_attention.py [hunk 4]
```diff
@@ -191,6 +194,52 @@ def score_mod(score, b, h, m, n):
 
         self.run_test(score_mod)
 
+    @supported_platform
+    def test_logsumexp_correctness(self):
```
Comment (Chillee) [inline test/inductor/test_templated_attention.py:198]: I would also test versions that only returns one of the the output buffers

File: torch/_dynamo/variables/higher_order_ops.py [hunk 5]
```diff
@@ -1400,13 +1400,8 @@ def create_scalar():
         lifted_args = tuple(arg for arg in body_lifted_freevars.keys())
 
         proxy_args = (body_node,) + lifted_args
-        example_value = pytree.tree_map_only(
```
Comment (Chillee) [inline torch/_dynamo/variables/higher_order_ops.py:1403]: can you roll the dtype changes into your PR btw?

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -4668,13 +4687,15 @@ def get_inputs_that_alias_output(self):
         return [i.get_name() for i in self.inputs]
 
 
-def mark_node_as_mutating(cur_buffer, *mutated_ops):
+def mark_node_as_mutating(cur_buffer, *mutated_ops: IRNode):
```
Comment (Chillee) [inline torch/_inductor/ir.py:4690]: is this a correct signature? Since it's an * arg?

File: torch/_inductor/kernel/templated_attention.py [hunk 7]
```diff
@@ -256,6 +258,14 @@ def create_placeholder(name: str, dtype: torch.dtype) -> InputBuffer:
                 query.get_size(),
                 make_contiguous_strides_for(query.get_size()),
             )
+            # see NOTE:[TritonTemplates with multiple outputs]
+            logsumexp_shape = query.get_size()[:-1]  # [B, H, M]
+            logsumexp = empty_strided(
+                logsumexp_shape,
+                None,
+                dtype=query.get_dtype(),
```
Comment (Chillee) [inline torch/_inductor/kernel/templated_attention.py:266]: Is this right? It's not always a fp32 value or something?

**Review by Chillee — APPROVED** [review]
LGTM!

File: torch/_inductor/select_algorithm.py [hunk 8]
```diff
@@ -649,6 +666,7 @@ def make_kernel_render(out_node):
                 "allow_tf32": str(kwargs.get("ALLOW_TF32", None)),
                 "acc_type": str(kwargs.get("ACC_TYPE", None)),
             },
+            mutated_inputs=mutated_inputs,
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:669]: Could we add a test that we don't fuse any use of the mutated input into the templated attention ?

Could we also check that we are correctly cloning the input mutated args here in benchmark ?

Comment (eellison) [inline torch/_inductor/select_algorithm.py:669]: Could we add a test that we don't fuse any use of the mutated input into the templated attention ?

Could we also check that we are correctly cloning the input mutated args here in benchmark ?

File: torch/_dynamo/variables/higher_order_ops.py (hunk 5 above)
Comment (drisspg) [reply torch/_dynamo/variables/higher_order_ops.py:1403]: changed the scalars to float32

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (Chillee) [reply torch/_inductor/select_algorithm.py:669]: To be clear, no other kernels are actually being fused into templated attention. And the input buffer that's being mutated is only being created for this kernel specifically.

We're only marking it as "mutated" because triton templates (and in general, Triton kernels) do not directly support multiple outputs that are being allocated. So instead,  we say it has "one output, and two 'inputs' that are mutated". 

@drisspg This would be a good thing to make clear as a comment.

Comment (drisspg) [reply torch/_inductor/select_algorithm.py:669]: I left a note here: https://github.com/pytorch/pytorch/pull/124308/files#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134acR3646

Do you think I should move it, I also added a see note above but could move it down here

File: torch/_inductor/kernel/templated_attention.py (hunk 3 above)
Comment (drisspg) [reply torch/_inductor/kernel/templated_attention.py:163]: added the flag but inductor has no autograd information at this point so need to think about the proper way to implement this. The writes didnt lead to much slow down, but will incur extra memory that will likely be deallocated quickly in compile since there will be no uses. Still a spike though

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (Chillee) [reply torch/_inductor/select_algorithm.py:669]: You can probably also add a note in the FlashAttention lowering

Comment (drisspg) [reply torch/_inductor/select_algorithm.py:669]: Spoke offline, added assert into TritontTemplateBuffer to ensure that only tempaltedattention is using this for now

And added some fileecheck tests

File: test/inductor/test_templated_attention.py (hunk 4 above)
Comment (drisspg) [reply test/inductor/test_templated_attention.py]: added tests that return both out, lse | just LSE and our existing tests cover the just out case
